arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java

```diff
@@ -508,6 +433,179 @@ public static VectorizedReader<?> replaceWithMetadataReader(
     return reader;
   }
 
+  private final class VectorizedArrowReaderLogicalTypeAnnotationVisitor
+      implements LogicalTypeAnnotation.LogicalTypeAnnotationVisitor<Object> {
+    private final Field arrowField;
+    private final PrimitiveType primitive;
+
+    VectorizedArrowReaderLogicalTypeAnnotationVisitor(Field arrowField, PrimitiveType primitive) {
+      this.arrowField = arrowField;
+      this.primitive = primitive;
+    }
+
+    @Override
+    public Optional<Object> visit(
+        LogicalTypeAnnotation.StringLogicalTypeAnnotation stringLogicalType) {
+      return visitEnumJsonBsonString();
+    }
+
+    @Override
+    public Optional<Object> visit(LogicalTypeAnnotation.EnumLogicalTypeAnnotation enumLogicalType) {
+      return visitEnumJsonBsonString();
+    }
+
+    @Override
+    public Optional<Object> visit(LogicalTypeAnnotation.UUIDLogicalTypeAnnotation uuidLogicalType) {
+      // Fallback to allocation based on primitive type name
+      VectorizedArrowReader.this.allocateVectorBasedOnTypeName(
+          VectorizedArrowReader.this.columnDescriptor.getPrimitiveType(), arrowField);
+      return Optional.empty();
+    }
+
+    @Override
+    public Optional<Object> visit(
+        LogicalTypeAnnotation.DecimalLogicalTypeAnnotation decimalLogicalType) {
+      VectorizedArrowReader.this.vec =
+          arrowField.createVector(VectorizedArrowReader.this.rootAlloc);
+      switch (primitive.getPrimitiveTypeName()) {
+        case BINARY:
+        case FIXED_LEN_BYTE_ARRAY:
+          ((FixedSizeBinaryVector) VectorizedArrowReader.this.vec)
+              .allocateNew(VectorizedArrowReader.this.batchSize);
+          VectorizedArrowReader.this.readType = ReadType.FIXED_LENGTH_DECIMAL;
+          VectorizedArrowReader.this.typeWidth = primitive.getTypeLength();
+          break;
+        case INT64:
+          ((BigIntVector) VectorizedArrowReader.this.vec)
+              .allocateNew(VectorizedArrowReader.this.batchSize);
+          VectorizedArrowReader.this.readType = ReadType.LONG_BACKED_DECIMAL;
+          VectorizedArrowReader.this.typeWidth = (int) BigIntVector.TYPE_WIDTH;
+          break;
+        case INT32:
+          ((IntVector) VectorizedArrowReader.this.vec)
+              .allocateNew(VectorizedArrowReader.this.batchSize);
+          VectorizedArrowReader.this.readType = ReadType.INT_BACKED_DECIMAL;
+          VectorizedArrowReader.this.typeWidth = (int) IntVector.TYPE_WIDTH;
+          break;
+        default:
+          throw new UnsupportedOperationException(
+              "Unsupported base type for decimal: " + primitive.getPrimitiveTypeName());
+      }
```

Comment:
nit: newlines please after the end of blocks. See: https://iceberg.apache.org/contribute/#block-spacing